orttraining/orttraining/python/ort_trainer.py

```diff
@@ -691,6 +693,73 @@ def load_state_dict(self, state_dict, strict=False):
+
+        if self.partition_optimizer_:
+            checkpoint_name = checkpoint_prefix + "_ZeROrank_" + str(self.world_rank) + "_of_" + str(self.world_size-1) + ".tar"
+        else:
+            checkpoint_name = checkpoint_prefix + ".tar" 
+        checkpoint_file = os.path.join(checkpoint_dir, checkpoint_name)
+
+        if os.path.exists(checkpoint_file):
+            print("WARNING: {} already exists, overwriting.".format(checkpoint_file))
+
+        print("Saving checkpoint {}".format(checkpoint_file))
+        torch.save(checkpoint_state_dict, checkpoint_file)
+        print("Saved checkpoint at: {}".format(checkpoint_file))
+    
+    def _load_single_checkpoint(self, checkpoint_dir, checkpoint_prefix, is_partitioned, strict):
+
+        if is_partitioned:
+            checkpoint_file = os.path.join(checkpoint_dir, checkpoint_prefix + "_ZeROrank_" + str(self.world_rank) + "_of_" + str(self.world_size-1) + ".tar")
+            assert_msg = ("Couldn't find checkpoint file {}." +
+                "Optimizer partitioning is enabled using ZeRO. Please make sure that the "+
+                "checkpoint file exists for rank {} of {}.").format(checkpoint_file,self.world_rank, self.world_size)
+            assert os.path.exists(checkpoint_file), assert_msg
+        else:
+            checkpoint_file = os.path.join(checkpoint_dir, checkpoint_prefix + ".tar")
+            assert_msg = "Couldn't find checkpoint file {}.".format(checkpoint_file)
+            assert os.path.exists(checkpoint_file), assert_msg
+
+        print("Loading checkpoint {}".format(checkpoint_file))
+        checkpoint_state = torch.load(checkpoint_file, map_location='cpu')
+
+        self.load_state_dict(checkpoint_state['model'], strict=strict)
+        del(checkpoint_state['model'])
+        return checkpoint_state
+
+    def _load_multi_checkpoint(self, checkpoint_dir, checkpoint_prefix, strict):
+        checkpoint_files = list_checkpoint_files(checkpoint_dir, checkpoint_prefix)
+        assert len(checkpoint_files) > 0, "No checkpoint files found with prefix \"{}\" in directory {}.".format(checkpoint_prefix, checkpoint_dir)
+        print("Loading checkpoint files: {}".format(checkpoint_files))
+
+        ckpt_agg = Combine_Zero_Checkpoint(checkpoint_files)
+        aggregate_state_dict = ckpt_agg.aggregate_checkpoints()
+
+        self.load_state_dict(aggregate_state_dict, strict=strict)
+
+        # aggregate other keys in the state_dict. 
+        # Values will be overwritten for matching keys among workers
+        all_checkpoint_states=dict()
+        for checkpoint_file in checkpoint_files:
+            checkpoint_state = torch.load(checkpoint_file, map_location='cpu')
+            del(checkpoint_state['model'])
+            all_checkpoint_states.update(checkpoint_state)            
+        return all_checkpoint_states
+
+    def load_checkpoint(self, checkpoint_dir, checkpoint_prefix="ORT_checkpoint", is_partitioned=False, strict=False):
```

Comment:
There can be a scenario where there are both of these in the same directory:
bert_toy_lamb.ZeROrank.1.3.ort.pt
bert_toy_lamb.ort.pt
In that case it would be unclear which checkpoint the user is referring to.

What we could do is, if the user forgets to pass the `is_partitioned` flag and no checkpoint `bert_toy_lamb.ort.pt ` exists, we could still look if zero checkpoints exist in the directory.

Reply:
The more parameters we have on the API, more scenarios we have to cover in unit tests as we have to test everything we publish. More confused the user can be, as other frameworks doesnt use such flags.

What if we do the simplest thing and clearly document it? Something like "load_checkpoint(checkpoint_dir) must point to an existing directory with a single checkpoint in it".

This way we dont have to create folders (which open space for security issues) and we dont have to care about multiple checkpoints in the same folder (the best practice would be using one checkpoint per folder, especially for zero which handles many files for a single checkpoint).

It is noble to try covering all ways users can organize their stuff, but it will double the size of source code, make mainteinance hard, require more tests and still we will never cover everything the user can do. Keeping it simple is usually the way to go :)

Reply:
In order to not add `is_partitioned` to the API surface, enforcing that the checkpoint folder must not contain multiple checkpoint files unidentified by the prefix, unless it is a partitioned checkpoint.